spacy/tests/matcher/test_matcher_logic.py

```diff
@@ -786,3 +798,63 @@ def test_matcher_with_alignments_non_greedy(en_vocab):
         for _, s, e, expected in matches:
             assert expected in results, (case_id, string, pattern_str, s, e, n_matches)
             assert len(expected) == e - s
+
+
+def test_matcher_non_greedy_operator(en_vocab):
+    # fmt: off
+    cases = [
+        (0, "aabbab", "a*? b", ["a a b", "a b", "b", "b", "a b", "b"]),
+        (1, "aabbab", "a+? b", ["a b", "a a b", "a b"]),
+        (2, "aabbab", "a*? b+?", ["a a b", "a b", "b", "b", "a b", "b"]),
+        (3, "aabbab", "a*? b*?", []),
+        (4, "aabbab", "b*? b*?", []),
+        (5, "aabbab", "b+? b+?", ["b b"]),
+        (6, "aabbab", "a* b*?", ["a", "a a", "a", "a"]),
+        (7, "aabbab", "a*? b*", ["a a b", "a b", "b", "a a b b", "a b b", "b b", "b", "a b", "b"]),
+        (8, "aabbc", "a* b*? c*?", ["a", "a a", "a"]),
+        (9, "aabbc", "a* b*? c", ["a a b b c", "a b b c", "b b c", "b c", "c"]),
+        (10, "abc", "a* b*? c*", ["a", "a b c", "b c", "c"]),
```

Comment:
I'm a little confused by this test case (with four matches) vs. the comments (with three matches). I'd (naively) expect to also see a match for `"a b"` if `"b c"` is also a match?

Reply:
Wait, I think might have backwards. If so, why is `"b c"` also a match?

Reply:
Hi there, apologies for the delay, was quite preoccupied with finals and only had the time to reply now.

spaCy matcher's behaviour is slightly different from regex where the spaCy `*` implementation returns **all possible matches** instead of regex `*` returning only 1 match.
### Case of `a*b*c*` 
#### Regex matching
![Screenshot 2022-12-09 at 11 46 55 PM](https://user-images.githubusercontent.com/82353723/206739943-dbed356a-cc1f-4c3e-bfd0-7b56749b215e.png)

For example in the case of the pattern `a*b*c*` on the text `"abc"`, regular expression will return only abc as seen in the screenshot above.

#### spaCy matching
![Screenshot 2022-12-08 at 1 50 31 AM](https://user-images.githubusercontent.com/82353723/206257951-aa2a4a6d-baa7-4adf-98e5-0387a3eef235.png)
As for spaCy matcher, if we were to apply `pattern = [{'ORTH': 'a', 'OP': '*'}, {'ORTH': 'b', 'OP': '*'},{'ORTH': 'c', 'OP': '*'}]` onto ` a b c` as seen in the screenshot above, it returns **all possible matches** in contrast to regular regex.

### Case of `a*b*?c*` in my implementation
spaCy matcher will transverse over the tokens `"a b c"` with matching pattern `a*b*?c*.

Spacy should seek to return the following
```
a* | b*? | c* 
0  | ... | 0
1  | ... | 0
0  | ... | 1
1  | ... | 1
```
Where `...` is the shortest possible match, 0 is absent and 1 is present.
In the below cases, `x` represents cases where the result is removed as it did not fulfil the shortest possible match criteria.
#### matcher machine at position `"a"`
```
a b c
^
```
```
a* | b*? | c* 
0  |  0  | 0
1  |  0  | 0 --> return "a"
1  |  x  | 0 --> branch "terminated"
1  |  1  | 1 --> return "a b c"
```

#### matcher machine at position "b"
```
a b c
  ^
```
```
b*? | c* 
0   | 0
x   | 0
1   | 1 --> return "b c"
```
#### matcher machine at position "c"
```
a b c
    ^
```
```
c* 
0
1 --> return "c"
```

Hope this helps! Thanks!!!